### ISSUE TITLE
Linker error: multiple definition in OpenSSL.Objects

### ISSUE THREAD

**Issue by mightybyte** [phonohawk/HsOpenSSL#50]
I'm getting the following build error when building with GHC 7.10.2 on ubuntu 12.04.5 LTS (the distro I get on CircleCI).  Anyone have any ideas?

    Building HsOpenSSL-0.11.1.1...
    Preprocessing library HsOpenSSL-0.11.1.1...
    [ 1 of 33] Compiling OpenSSL.SSL.Option ( .stack-work/dist/x86_64-linux/Cabal-1.22.4.0/build/OpenSSL/SSL/Option.hs, .stack-work/dist/x86_64-linux/Cabal-1.22.4.0/build/OpenSSL/SSL/Option.o )
    [ 2 of 33] Compiling OpenSSL.ERR      ( OpenSSL/ERR.hs, .stack-work/dist/x86_64-linux/Cabal-1.22.4.0/build/OpenSSL/ERR.o )
    [ 3 of 33] Compiling OpenSSL.Stack    ( OpenSSL/Stack.hs, .stack-work/dist/x86_64-linux/Cabal-1.22.4.0/build/OpenSSL/Stack.o )
    [ 4 of 33] Compiling OpenSSL.DH.Internal ( OpenSSL/DH/Internal.hs, .stack-work/dist/x86_64-linux/Cabal-1.22.4.0/build/OpenSSL/DH/Internal.o )

    /tmp/stack108855/HsOpenSSL-0.11.1.1/OpenSSL/DH/Internal.hs:17:1: Warning:
        The import of ‘Control.Applicative’ is redundant
          except perhaps to import instances from ‘Control.Applicative’
        To import instances alone, use: import Control.Applicative()
    [ 5 of 33] Compiling OpenSSL.Objects  ( .stack-work/dist/x86_64-linux/Cabal-1.22.4.0/build/OpenSSL/Objects.hs, .stack-work/dist/x86_64-linux/Cabal-1.22.4.0/build/OpenSSL/Objects.o )
    /tmp/ghc109525_0/ghc_32.o: In function `r7OL_closure':
    (.data+0x0): multiple definition of `__stginit_HsOpezuGnLae7U7W9ZZI8AcRoM668H_OpenSSLziObjects'
    /tmp/ghc109525_0/ghc_32.o:(.data+0x0): first defined here
    /tmp/ghc109525_0/ghc_32.o: In function `r7pH_closure':
    (.data+0x10): multiple definition of `HsOpezuGnLae7U7W9ZZI8AcRoM668H_OpenSSLziObjects_getObjNames1_closure'
    /tmp/ghc109525_0/ghc_32.o:(.data+0x10): first defined here
    /tmp/ghc109525_0/ghc_32.o: In function `c7SA_info':
    (.text+0x2c8): multiple definition of `HsOpezuGnLae7U7W9ZZI8AcRoM668H_OpenSSLziObjects_getObjNames1_info'
    /tmp/ghc109525_0/ghc_32.o:(.text+0x2c8): first defined here
    /tmp/ghc109525_0/ghc_32.o: In function `r7pH_closure':
    (.data+0x20): multiple definition of `HsOpezuGnLae7U7W9ZZI8AcRoM668H_OpenSSLziObjects_getObjNames_closure'
    /tmp/ghc109525_0/ghc_32.o:(.data+0x20): first defined here
    /tmp/ghc109525_0/ghc_32.o: In function `c7W6_info':
    (.text+0xe98): multiple definition of `HsOpezuGnLae7U7W9ZZI8AcRoM668H_OpenSSLziObjects_getObjNames_info'
    /tmp/ghc109525_0/ghc_32.o:(.text+0xe98): first defined here
    /tmp/ghc109525_0/ghc_32.o: In function `r7pH_closure':
    (.data+0x30): multiple definition of `HsOpezuGnLae7U7W9ZZI8AcRoM668H_OpenSSLziObjects_MDMethodType_closure'
    /tmp/ghc109525_0/ghc_32.o:(.data+0x30): first defined here
    /tmp/ghc109525_0/ghc_32.o: In function `c7W6_info':
    (.text+0xeb8): multiple definition of `HsOpezuGnLae7U7W9ZZI8AcRoM668H_OpenSSLziObjects_MDMethodType_static_info'
    /tmp/ghc109525_0/ghc_32.o:(.text+0xeb8): first defined here
    /tmp/ghc109525_0/ghc_32.o: In function `r7pH_closure':
    (.data+0x38): multiple definition of `HsOpezuGnLae7U7W9ZZI8AcRoM668H_OpenSSLziObjects_CipherMethodType_closure'
    /tmp/ghc109525_0/ghc_32.o:(.data+0x38): first defined here
    /tmp/ghc109525_0/ghc_32.o: In function `c7W6_info':
    (.text+0xed8): multiple definition of `HsOpezuGnLae7U7W9ZZI8AcRoM668H_OpenSSLziObjects_CipherMethodType_static_info'
    /tmp/ghc109525_0/ghc_32.o:(.text+0xed8): first defined here
    /tmp/ghc109525_0/ghc_32.o: In function `r7pH_closure':
    (.data+0x40): multiple definition of `HsOpezuGnLae7U7W9ZZI8AcRoM668H_OpenSSLziObjects_PKeyMethodType_closure'
    /tmp/ghc109525_0/ghc_32.o:(.data+0x40): first defined here
    /tmp/ghc109525_0/ghc_32.o: In function `c7W6_info':
    (.text+0xef8): multiple definition of `HsOpezuGnLae7U7W9ZZI8AcRoM668H_OpenSSLziObjects_PKeyMethodType_static_info'
    /tmp/ghc109525_0/ghc_32.o:(.text+0xef8): first defined here
    /tmp/ghc109525_0/ghc_32.o: In function `r7pH_closure':
    (.data+0x48): multiple definition of `HsOpezuGnLae7U7W9ZZI8AcRoM668H_OpenSSLziObjects_CompMethodType_closure'
    /tmp/ghc109525_0/ghc_32.o:(.data+0x48): first defined here
    /tmp/ghc109525_0/ghc_32.o: In function `c7W6_info':
    (.text+0xf18): multiple definition of `HsOpezuGnLae7U7W9ZZI8AcRoM668H_OpenSSLziObjects_CompMethodType_static_info'
    /tmp/ghc109525_0/ghc_32.o:(.text+0xf18): first defined here
    /tmp/ghc109525_0/ghc_32.o: In function `r7pH_closure':
    (.data+0x50): multiple definition of `HsOpezuGnLae7U7W9ZZI8AcRoM668H_OpenSSLziObjects_ObjNameType_closure_tbl'
    /tmp/ghc109525_0/ghc_32.o:(.data+0x50): first defined here
    /tmp/ghc109525_0/ghc_31.o: In function `OpenSSLziObjects_d7Ce':
    ghc_29.c:(.text+0x0): multiple definition of `OpenSSLziObjects_d7Ce'
    /tmp/ghc109525_0/ghc_31.o:ghc_29.c:(.text+0x0): first defined here
    collect2: error: ld returned 1 exit status

**Comment by mightybyte** [comment]
Ok, I finally solved this issue.  It's the same as the one described here:

http://stackoverflow.com/questions/34654262/build-fails-because-of-multiple-definition-linker-errors-in-native-dependencie

It seems that the problem has to do with gcc-4.9.  The solution that worked for me was to drop back to an older gcc version.  In my specific case with CircleCI I did that as described here:

https://github.com/Yuras/Hate/blob/link-err/circle.yml#L3-L6

Hopefully having this information here can help someone else save time if they encounter this problem.
